### PR TITLE
bpf, maps: Don't set nodeid to 0 when nodeid allocation fails.

### PR DESCRIPTION
**Pull request by marseel**
Don't set nodeid to 0 when nodeid allocation fails.

When we run out of IDs to allocate for nodes, we returned zero value before that was later on propagated to bpf map.

```release-note
Fixed issue with assigning 0 nodeID when corresponding bpf map run out of space. 
Potentially it could have impacted connectivity in large clusters (>4k nodes) with IPSec or Mutual Auth enabled.
Otherwise, it was merely generating unnecessary error log messages.
```

### CODE REVIEW

File: pkg/datapath/linux/node_ids.go [hunk 1]
```diff
@@ -92,6 +92,12 @@ func (n *linuxNodeHandler) allocateIDForNode(node *nodeTypes.Node) (uint16, erro
 		}
 	}
 
+	// If we failed to allocate nodeID, don't map any IP to 0 nodeID.
+	// This causes later errors like "Found a foreign IP address with the ID of the current node"
+	if nodeID == uint16(idpool.NoID) {
+		return nodeID, errs
+	}
```
Comment (pchaigno) [inline pkg/datapath/linux/node_ids.go]: Can't we simply return on line 86? I think I just forgot the return statement there.

File: pkg/datapath/linux/node_ids.go [hunk 2]
```diff
@@ -92,6 +92,12 @@ func (n *linuxNodeHandler) allocateIDForNode(node *nodeTypes.Node) (uint16, erro
 		}
```
Comment (pchaigno) [inline pkg/datapath/linux/node_ids.go:93]: I'm not able to parse the description for the first commit. I get what the commit's intent is, but might be worth rephrasing for others.

File: pkg/datapath/linux/node_ids.go [hunk 3]
```diff
@@ -280,6 +284,13 @@ func (n *linuxNodeHandler) RestoreNodeIDs() {
 	}
 
 	n.registerNodeIDAllocations(nodeIDs)
+	for ip := range incorrectNodeIDs {
+		if err := n.unmapNodeID(ip); err != nil {
+			log.WithError(err).WithFields(logrus.Fields{
+				logfields.IPAddr: ip,
+			}).Warn("Failed to remove a incorrect node IP to node ID mapping")
+		}
+	}
```
Comment (pchaigno) [inline pkg/datapath/linux/node_ids.go:293]: Should we also print a warning about the fact we found such incorrect mappings? I understand right now we expect some users have accumulated hundreds of incorrect mappings, but longer term we won't be expecting any.

File: pkg/datapath/linux/node_ids.go [hunk 4]
```diff
@@ -274,6 +284,13 @@ func (n *linuxNodeHandler) RestoreNodeIDs() {
...
 	log.Infof("Restored %d node IDs from the BPF map", len(nodeIDs))
 }
```
Comment (pchaigno) [inline pkg/datapath/linux/node_ids.go:296]: (*Commenting in random place.*)

Could we describe the user impact in the release note? AFAICS, without Mutual Auth and IPsec, the only impact are superfluous error messages in logs. With IPsec, the failure to allocate node IDs would impact connectivity to new nodes in the cluster(mesh). With Mutual Auth, I'm unsure of the impact. Probably easier to describe each situation on a new line.

**Review by pchaigno — COMMENTED** [review]
Thanks for fixing this Marcel! :pray:

File: pkg/datapath/linux/node_ids.go (hunk 2 above)
Comment (marseel) [reply pkg/datapath/linux/node_ids.go:93]: I changed it, lmk if it looks better.

File: pkg/datapath/linux/node_ids.go (hunk 3 above)
Comment (marseel) [reply pkg/datapath/linux/node_ids.go:293]: done.

File: pkg/datapath/linux/node_ids.go [hunk 5]
```diff
@@ -274,6 +281,16 @@ func (n *linuxNodeHandler) RestoreNodeIDs() {
 	}
 
 	n.registerNodeIDAllocations(nodeIDs)
+	if len(incorrectNodeIDs) > 0 {
+		log.WithFields(logrus.Fields{}).Warnf("Removing %d incorrect node IP to node ID mappings from the BPF map", len(incorrectNodeIDs))
```
Comment (pchaigno) [inline pkg/datapath/linux/node_ids.go]:
```suggestion
		log.Warnf("Removing %d incorrect node IP to node ID mappings from the BPF map", len(incorrectNodeIDs))
```
or am I missing something?

**Review by pchaigno — APPROVED** [review]

File: pkg/datapath/linux/node_ids.go (hunk 5 above)
Comment (marseel) [reply pkg/datapath/linux/node_ids.go]: fixed

**Review by tommyp1ckles — APPROVED** [review]

**Comment by gandro** [comment]
The error handling of `allocateIDForNode` works slightly differently in v1.14 and older. Since I don't have any context and this fix seems rather subtle, I'm deferring the backport to those branches to the authors.

**Pull request merged**